### ISSUE TITLE
`kitty +kitten ssh` broken

### ISSUE THREAD

**Issue by WXZhao7** [kovidgoyal/kitty#5652]
**Describe the bug**
`kitty +kitten ssh padavan` exit after I input the password of private key. But either `ssh padavan` or `kitty +kitten ssh other-host` worked.

This is my ~/.ssh/conf
```ssh
Host padavan
    HostName 000.000.000.000
    User admin
    Port 12345
    PreferredAuthentications publickey
    IdentityFile /home/wxz/.ssh/id_rsa_padavan
```

**To Reproduce**
1. XXX
2. YYY
3. ZZZ

**Screenshots**

**Environment details**
```
kitty 0.26.5 created by Kovid Goyal
Linux WXZ-Ubuntu20 5.15.0-52-generic #58~20.04.1-Ubuntu SMP Thu Oct 13 13:09:46 UTC 2022 x86_64
Ubuntu 20.04.5 LTS WXZ-Ubuntu20 /dev/tty

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.5 LTS"
Running under: X11
Frozen: True
Paths:
  kitty: /home/wxz/.local/kitty.app/bin/kitty
  base dir: /home/wxz/.local/kitty.app/lib/kitty
  extensions dir: /home/wxz/.local/kitty.app/lib/kitty-extensions
  system shell: /bin/zsh

Config options different from defaults:

Important environment variables seen by the kitty process:
	PATH                                /home/wxz/Program/miniconda3/envs/seismic/bin:/home/wxz/Program/miniconda3/condabin:/opt/TauP-2.5.1/bin:/opt/IASP91/iasp91:/opt/GMT/GMT-4.5.18/bin:/opt/GMT/GMT-6.4.0/bin:/home/wxz/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/wxz/.local/share/JetBrains/Toolbox/scripts:/usr/local/sac/bin
	LANG                                en_US.UTF-8
	SHELL                               /bin/zsh
	DISPLAY                             :1
	USER                                wxz
	LC_TIME                             en_ZM.UTF-8
	XDG_SESSION_TYPE                    x11
	LC_MONETARY                         en_ZM.UTF-8
	XDG_SESSION_CLASS                   user
	LC_ADDRESS                          en_ZM.UTF-8
	XDG_RUNTIME_DIR                     /run/user/1000
	XDG_MENU_PREFIX                     gnome-
	LC_TELEPHONE                        en_ZM.UTF-8
	XDG_CURRENT_DESKTOP                 ubuntu:GNOME
	XDG_SESSION_DESKTOP                 ubuntu
	LC_NAME                             en_ZM.UTF-8
	LC_MEASUREMENT                      en_ZM.UTF-8
	LC_IDENTIFICATION                   en_ZM.UTF-8
	XDG_CONFIG_DIRS                     /etc/xdg/xdg-ubuntu:/etc/xdg
	XDG_DATA_DIRS                       /usr/share/ubuntu:/usr/local/share/:/usr/share/
	LC_NUMERIC                          en_ZM.UTF-8
	LC_PAPER                            en_ZM.UTF-8

```

**Additional context**
Try to reproduce the problem with `kitty --config NONE` if you cannot then post a minimal kitty.conf that reproduces the problem. If the problem involves interaction with some other terminal program post a minimal config for that program to reproduce the problem as well.

```
❯ ksh -v padavan
OpenSSH_8.2p1 Ubuntu-4ubuntu0.5, OpenSSL 1.1.1f  31 Mar 2020
debug1: Reading configuration data /home/wxz/.ssh/config
debug1: /home/wxz/.ssh/config line 33: Applying options for padavan
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: auto-mux: Trying existing master
debug1: Control socket "/run/user/1000/kssh-1951054-7cd9d94a494282f231fa11b558a44168fabc3ca0" does not exist
debug1: Connecting to 000.000.000.000 [000.000.000.000] port 12345.
debug1: Connection established.
debug1: identity file /home/wxz/.ssh/id_rsa_padavan type -1
debug1: identity file /home/wxz/.ssh/id_rsa_padavan-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.5
debug1: Remote protocol version 2.0, remote software version dropbear_2018.76
debug1: no match: dropbear_2018.76
debug1: Authenticating to 000.000.000.000:12345 as 'admin'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: ecdh-sha2-nistp256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: aes128-ctr MAC: hmac-sha2-256 compression: none
debug1: kex: client->server cipher: aes128-ctr MAC: hmac-sha2-256 compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:h6HRaQY/Jj9f8DdoGm8m+AxKUho/Ma9dOVSnBQNMeDk
debug1: checking without port identifier
debug1: Host '000.000.000.000' is known and matches the ECDSA host key.
debug1: Found key in /home/wxz/.ssh/known_hosts:16
debug1: found matching key w/out port
debug1: rekey out after 4294967296 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 4294967296 blocks
debug1: Will attempt key: /home/wxz/.config/gsconnect/private.pem RSA SHA256:geUlwVgmqMV/TXBQaVB2f2v0XhhDKD4hHDnq0jG/w+k agent
debug1: Will attempt key: wxzhao@qq.com RSA SHA256:i0bVRKytCngne5MS1vmpbrWH+EHJqIEDpXk5EaDDAxA agent
debug1: Will attempt key: wxz@WXZ-Ubuntu20 RSA SHA256:1cZZGqTH8hMUOpxrqDnYWthpXafhdw5uxd+r3Kvq7OU agent
debug1: Will attempt key: wxzhao7@gmail.com RSA SHA256:6mgb2NQT61THT1efVaec4UgHM2YproxgDLAJ//upMXw agent
debug1: Will attempt key: /home/wxz/.ssh/id_rsa_padavan  explicit
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: /home/wxz/.config/gsconnect/private.pem RSA SHA256:geUlwVgmqMV/TXBQaVB2f2v0XhhDKD4hHDnq0jG/w+k agent
debug1: Authentications that can continue: publickey
debug1: Offering public key: wxzhao@qq.com RSA SHA256:i0bVRKytCngne5MS1vmpbrWH+EHJqIEDpXk5EaDDAxA agent
debug1: Authentications that can continue: publickey
debug1: Offering public key: wxz@WXZ-Ubuntu20 RSA SHA256:1cZZGqTH8hMUOpxrqDnYWthpXafhdw5uxd+r3Kvq7OU agent
debug1: Authentications that can continue: publickey
debug1: Offering public key: wxzhao7@gmail.com RSA SHA256:6mgb2NQT61THT1efVaec4UgHM2YproxgDLAJ//upMXw agent
debug1: Authentications that can continue: publickey
debug1: Trying private key: /home/wxz/.ssh/id_rsa_padavan
Enter passphrase for key '/home/wxz/.ssh/id_rsa_padavan': 
debug1: Authentication succeeded (publickey).
Authenticated to 000.000.000.000 ([000.000.000.000]:12345).
debug1: setting up multiplex master socket
debug1: channel 0: new [/run/user/1000/kssh-1951054-7cd9d94a494282f231fa11b558a44168fabc3ca0]
debug1: control_persist_detach: backgrounding master process
debug1: forking to background
debug1: Entering interactive session.
debug1: pledge: id
debug1: multiplexing control connection
debug1: channel 1: new [mux-control]
debug1: channel 2: new [client-session]
debug1: Sending environment.
debug1: Sending env LC_TIME = en_ZM.UTF-8
debug1: Sending env LC_MONETARY = en_ZM.UTF-8
debug1: Sending env LC_ADDRESS = en_ZM.UTF-8
debug1: Sending env LANG = en_US.UTF-8
debug1: Sending env LC_TELEPHONE = en_ZM.UTF-8
debug1: Sending env LC_NAME = en_ZM.UTF-8
debug1: Sending env LC_MEASUREMENT = en_ZM.UTF-8
debug1: Sending env LC_IDENTIFICATION = en_ZM.UTF-8
debug1: Sending env LC_NUMERIC = en_ZM.UTF-8
debug1: Sending env LC_PAPER = en_ZM.UTF-8
    if [ -z "$python" -o ython=$(command -v python))*" > /dev/stderr; cleanup_on_bootstrap_exit; exit 1; }4\\\\\\n\\\\\\041)"'  '/bin/sh
{ Distributed under terms of the GPLv3 license.dgoyal.net>
  unalias command; 
die() { printf " ] && command rm -rf "$tdir"ho" 2> /dev/null < /dev/tty
                033[31m%s
                         033[m
                              n
    base64_encode() { command base64 | command tr -d turn 1; fi 1; fistrap_exit; exit 1; }

                                                     n

    base64_encode() { command b64encode - | command sed  }
                                                        1d;$d
                                                              | command tr -d 

                                                                              n

    pybase64() { command "$python" -c "import sys, base64; getattr(sys.stdout, r; }
                                                                               buffer
                                                                                     , sys.stdout).write(base64.standard_b64$1(getattr(sys.stdin, 
                                                                                                                                                  buffer
                                                                                                                                                        , sys.stdin).read()))"    base64_encode() { command "$perl" -MMIME::Base64 -0777 -ne 
                                                               print encode_base64($_)
    base64_decode() { command "$perl" -MMIME::Base64 -ne                              ; }
                                                         print decode_base64($_)
dcs_to_kitty() { printf "e not present on remote host, ssh kitten cannot function."
                         033P@kitty-$1|%s
                                         033
    dcs_to_kitty "ssh" "id="1951054-1":pwfile="/kssh-1951264-8e45dc73785d2f25671ea0a093596760c3532309d1f0a3c504021c39de4bba2f":pw="1ad64501b2f5002f42b31d1aba23b856d2c6e0ab113: eval: line 26: command: not foundecute it here"thenoot" ""ad SSH data from tty" /dev/nulle in the futureequires tar."
{ Distributed under terms of the GPLv3 license.dgoyal.net>
  unalias command; 
die() { printf " ] && command rm -rf "$tdir"ho" 2> /dev/null < /dev/tty
                033[31m%s
                         033[m
                              n
    base64_encode() { command base64 | command tr -d turn 1; fi 1; fistrap_exit; exit 1; }

                                                     n

    base64_encode() { command b64encode - | command sed  }
                                                        1d;$d
                                                              | command tr -d 

                                                                              n

    pybase64() { command "$python" -c "import sys, base64; getattr(sys.stdout, r; }
                                                                               buffer
                                                                                     , sys.stdout).write(base64.standard_b64$1(getattr(sys.stdin, 
                                                                                                                                                  buffer
                                                                                                                                                        , sys.stdin).read()))"    base64_encode() { command "$perl" -MMIME::Base64 -0777 -ne 
                                                               print encode_base64($_)
    base64_decode() { command "$perl" -MMIME::Base64 -ne                              ; }
                                                         print decode_base64($_)
dcs_to_kitty() { printf "e not present on remote host, ssh kitten cannot function."
                         033P@kitty-$1|%s
                                         033
    dcs_to_kitty "ssh" "id="1951054-1":pwfile="/kssh-1951264-8e45dc73785d2f25671ea0a093596760c3532309d1f0a3c504021c39de4bba2f":pw="1ad64501b2f5002f42b31d1aba23b856d2c6e0ab113: eval: line 27: command: not foundecute it here"thenoot" ""ad SSH data from tty" /dev/nulle in the futureequires tar."
{ Distributed under terms of the GPLv3 license.dgoyal.net>
  unalias command; 
die() { printf " ] && command rm -rf "$tdir"ho" 2> /dev/null < /dev/tty
                033[31m%s
                         033[m
                              n
    base64_encode() { command base64 | command tr -d turn 1; fi 1; fistrap_exit; exit 1; }

                                                     n

    base64_encode() { command b64encode - | command sed  }
                                                        1d;$d
                                                              | command tr -d 

                                                                              n

    pybase64() { command "$python" -c "import sys, base64; getattr(sys.stdout, r; }
                                                                               buffer
                                                                                     , sys.stdout).write(base64.standard_b64$1(getattr(sys.stdin, 
                                                                                                                                                  buffer
                                                                                                                                                        , sys.stdin).read()))"    base64_encode() { command "$perl" -MMIME::Base64 -0777 -ne 
                                                               print encode_base64($_)
    base64_decode() { command "$perl" -MMIME::Base64 -ne                              ; }
                                                         print decode_base64($_)
dcs_to_kitty() { printf "e not present on remote host, ssh kitten cannot function."
                         033P@kitty-$1|%s
                                         033
    dcs_to_kitty "ssh" "id="1951054-1":pwfile="/kssh-1951264-8e45dc73785d2f25671ea0a093596760c3532309d1f0a3c504021c39de4bba2f":pw="1ad64501b2f5002f42b31d1aba23b856d2c6e0ab113: eval: line 28: command: not foundecute it here"thenoot" ""ad SSH data from tty" /dev/nulle in the futureequires tar."
{ Distributed under terms of the GPLv3 license.dgoyal.net>
  unalias command; 
die() { printf " ] && command rm -rf "$tdir"ho" 2> /dev/null < /dev/tty
                033[31m%s
                         033[m
                              n
    base64_encode() { command base64 | command tr -d turn 1; fi 1; fistrap_exit; exit 1; }

                                                     n

    base64_encode() { command b64encode - | command sed  }
                                                        1d;$d
                                                              | command tr -d 

                                                                              n

    pybase64() { command "$python" -c "import sys, base64; getattr(sys.stdout, r; }
                                                                               buffer
                                                                                     , sys.stdout).write(base64.standard_b64$1(getattr(sys.stdin, 
                                                                                                                                                  buffer
                                                                                                                                                        , sys.stdin).read()))"    base64_encode() { command "$perl" -MMIME::Base64 -0777 -ne 
                                                               print encode_base64($_)
    base64_decode() { command "$perl" -MMIME::Base64 -ne                              ; }
                                                         print decode_base64($_)
dcs_to_kitty() { printf "e not present on remote host, ssh kitten cannot function."
                         033P@kitty-$1|%s
                                         033
    dcs_to_kitty "ssh" "id="1951054-1":pwfile="/kssh-1951264-8e45dc73785d2f25671ea0a093596760c3532309d1f0a3c504021c39de4bba2f":pw="1ad64501b2f5002f42b31d1aba23b856d2c6e0ab113: eval: line 40: command: not foundecute it here"thenoot" ""ad SSH data from tty" /dev/nulle in the futureequires tar."
debug1: client_input_channel_req: channel 2 rtype exit-status reply 0
debug1: channel 2: free: client-session, nchannels 3
debug1: channel 1: free: mux-control, nchannels 2
Shared connection to 000.000.000.000 closed.
```

**Comment by kovidgoyal** [comment]
I need a way to reproduce your problem, otherwise I have no way to help
you. If you can setup ssh access to a test host that reproduces the
issue, let me know and I will take a look.

**Comment by page-down** [comment]
The ssh kitten works with dropbear. I tested `dropbear 2018.76`, a four year old version, on a two year old linux system.
It looks like you are missing some required commands in your device, e.g. `tar`, `base64`, `stty`, etc.
There is a big chance that ssh kitten will not work properly on a stripped system.

Also in embedded devices with a limited number of writes, it may lead to flash memory wearout.
This is because ssh kitten packs and unpacks files on every connection, even if a shared connection has been used, and even if the files were updated to the latest a few seconds ago.
